Making of NumTok an API for numeral

herbelin opened the pull request:
**Kind:** code reorganization

Has an overlay for QuickChick: [#189](https://github.com/QuickChick/QuickChick/pull/189).

Depends on #11688 (merged).

Fixes at least one bug in the parser of decimal numbers for `R`.

We make an API out of the different pieces of code which manipulates numerals, possibly with a sign, a decimal part or an exponent. In particular, we use typing to distinguish when it is about unsigned numbers and when it is about signed numbers, and when it is about integer numbers.

Former type `NumTok.t` is now abstract (internally made of `string`s accepting leading `0`s, and  thousands separated by `_`).

The API is somehow constraining and (therefore) painful. One has to find in the API the right conversions between OCaml's `string` representation of `int`, `bigint`, Coq's abstract `NumTok.SignedNat` and `NumTok.UnsignedNat` representations of signed or unsigned int, Coq's abstract `NumTok.Signed` and `NumTok.Unsigned` representations of signed or unsigned int with decimals and exponents. On the other side, these constraints are probably useful.

- [X] Added / updated test-suite

proux01 commented on proofs/logic.ml:
```diff
@@ -67,6 +67,7 @@ let catchable_exception = function
   | Nametab.GlobalizationError _
   (* reduction errors *)
   | Tacred.ReductionTacticError _ -> true
+  | Logic_monad.TacticFailure _ -> true
```
I don't get this change, is it related?

herbelin replied:
Sorry, a collision with another PR I was working on at roughly the same time. Fixed

proux01 commented on interp/numTok.mli:
```diff
@@ -1,18 +1,88 @@
-val equal : t -> t -> bool
+(** Numerals in different forms: signed or unsigned, possibly with
+    fractional part and exponent; we preserve the original user
+    representation for better error reporting *)
```
If I remember well, it is also needed for user defined notations that would contain numerals (c.f. comment below)

https://github.com/coq/coq/blob/18aa9ca60ec9b3d1712276ec0c615dfe54c1a251/interp/constrexpr.ml#L64

herbelin replied:
I changed the message. Tell me if you consider that it fixes the previous inaccuracy.

proux01 replied:
I would rather move the entire comment from constrexpr.ml (which seems now a bit out of place / inaccurate) to numTok.mli.

proux01 replied:
BTW, shouldn't we take the opportunity to add a test for this, something along the following lines
```Coq
Notation "0" := Prop.
Notation "+0" := O.
Notation "-0" := true.
Notation "00" := (nil, nil).                                                                                                                          
Check 0 : Prop.
Check +0 : nat.
Check -0 : bool.
Check 00 : prod _ _.

Notation "1000" := False.
Notation "1_000" := O.
Check 1000 : Prop.
Check 1_000 : nat.
```

proux01 commented on interp/numTok.mli:
```diff
@@ -1,18 +1,88 @@
+
+  val of_decimal_and_exponent : SignedNat.t -> UnsignedNat.t option -> SignedNat.t option -> t
+  val to_decimal_and_exponent : t -> SignedNat.t * UnsignedNat.t option * SignedNat.t option
+    (* n, p and q such that the number is n.p*2^q *)
```
Isn't it `n.p*10^q`?

herbelin replied:
Right, and similarly below.

proux01 commented on interp/numTok.mli:
```diff
@@ -1,18 +1,88 @@
+
+  val to_bigint_fraction : t -> Bigint.bigint * Bigint.bigint
+  val of_bigint_fraction : Bigint.bigint -> Bigint.bigint -> t
+    (* n and p such that the number is n/2^p *)
```
`n/10^p`?

proux01 commented on interp/numTok.ml:
```diff
@@ -1,52 +1,219 @@
+  let of_string s =
+    assert (String.length s > 0);
+    assert (s.[0] <> '-');
+    String.(concat "" (split_on_char '_' s))
```
Doesn't this contradict the comment above?

herbelin replied:
Indeed, I made a confusion. The `split_on_char '_' s` is actually for the `NumTok.UnsignedNat.t` to `string` translation (where `string` is typically the input syntax for `int_of_string` or `Bigint.of_string).

Moreover, I realize that some of the usage of `NumTok.UnsignedNat.to_string` done outside `NumTok` are rather of the `print` feature. So, I added a `NumTok.UnsignedNat.print` method.

The last use of `NumTok.UnsignedNat.to_string` is in `notation.ml` and I cannot drop the specific treatment of the `_` separator there.

Thanks!

proux01 commented on interp/numTok.ml:
```diff
@@ -1,52 +1,219 @@
+
+  (** Comparing two raw numbers (base 10, big-endian, non-negative).
+      A bit nasty, but not critical: only used to decide when a
+      number is considered as large (see warnings above). *)
```
No more "warnings above"

herbelin replied:
I replaced with a reference to `notation.ml`.

proux01 commented on interp/numTok.ml:
```diff
@@ -1,52 +1,219 @@
+struct
+
+  type t = {
+    int : string;  (** -?\[0-9\]\[0-9_\]* *)
```
`-?` for unsigned numbers?

herbelin replied:
This is remaining from a previous step of the reworking of `NumTok`, I fixed.

proux01 commented on interp/numTok.ml:
```diff
@@ -1,52 +1,219 @@
+  let equal (s1,n1) (s2,n2) =
+    s1 = s2 && equal n1 n2
+
+  let is_zero s =
```
This can probably be implemented using `UnsignedNat.compare` above, saving a few lines.

herbelin replied:
Good idea, I did it. (Please check though.)

proux01 replied:
Which compare is this? (I don't see any compare in Unsigned)

herbelin replied:
Indeed. This was the wrong `compare` (from `Pervasives`) and, more than that, I completely broke the code. Hopefully fixed.

proux01 commented on interp/numTok.ml:
```diff
@@ -1,52 +1,219 @@
+    let sign, int = SignedNat.of_bigint n in
+    (sign, { int; frac = ""; exp = "" })
+
+  let to_bigint_fraction (s, { int; frac; exp }) =
```
Don't we need to remove the underscores (like in `to_bigint` just above)?

herbelin replied:
Yes, well spotted.

I'm going tonight to clarify the semantics of the different `to_string` functions.

proux01 commented:
The commit message reads "positive natural numbers (may include "_" to separate thousands and leading 0)" but strictly speaking, underscores are allowed everywhere.

herbelin commented:
> The commit message reads "positive natural numbers (may include "_" to separate thousands and leading 0)" but strictly speaking, underscores are allowed everywhere.

Good question. I'm unsure that having `_` is actually very intuitive and I'm wondering whether we should not be more restrictive. I don't know the general feeling.

herbelin commented:
@proux01: Actually, no need you look now at the change. I'll make tonight a clear diff of the new changes easier to read.

herbelin commented:
@proux01: Thanks a lot for your careful reading then.

I reorganize a bit the `of_string`/`to_string` according to your remarks. I added more comments in `numTok.mli`. There are now 3 new commits compared to what you read in the morning (one of them to be squashed with the main commit before merging).

The logic for the changes is:
- `of_string`/`to_string` go from OCaml's `int_of_string` or `Bigint.of_string` string format in the `SignedNat`/`UnsignedNat` cases. For OCaml, I realized that `_` is actually supported. We need to remove the `_` for `Bigint.of_string` though (even if `bigint` may eventually disappear in #8743). So maybe overkill to skip `_`, but, at least safer.
- `of_string`/`to_string` goes from OCaml `float_of_string` format in the full `Signed`/`Unsigned` cases.
- new `sprint`/`print` to restore the string as it was parsed, in all types

I have further questions if ever you have an opinion on them:
- Should we parse `+42`, or `+3.5`, the same way as OCaml actually accepts to parse it? This seems quite standard to parse positive s after all, so, maybe should we? I experimentally added a commit for supporting it (in `NumTok`, not yet in the parse).
- Should I complete the API for consistency? For instance, I currently added, even if unused: `of_string`/`to_string`/`of_bigint` in `SignedNat` and `sprint`/`compare` in `UnsignedNat`. I may complete further, e.g. `sprint`/`print` in all categories??
- Should I rename `Signed`/`Unsigned` into `SignedDecimal`/`UnsignedDecimal` to make a stronger difference?

proux01 commented on interp/numTok.ml:
```diff
@@ -201,12 +225,12 @@ struct
+      | '+' -> Bigint.of_string (UnsignedNat.to_string (String.sub exp 2 (String.length exp - 2)))
+      | '-' -> Bigint.(neg (of_string (UnsignedNat.to_string (String.sub exp 2 (String.length exp - 2)))))
+      | _ -> Bigint.of_string (UnsignedNat.to_string (String.sub exp 1 (String.length exp - 1))) in
       Bigint.(sub e (of_int (String.length frac))) in
```
Don't we have to remove potential `_` from `frac` (but not leading zeros)? `UnsignedNat.to_string frac` should do the job.

herbelin replied:
Indeed, this is subtle (and was wrong before). The length needs to be done once `_` are removed.

Here is a bug of 8.11:
```
Require Import Reals.
Check 1.5_%R.
(* 15e-2%R *)
Check 1.5_e2%R.
(* 15%R *)
```

Incidentally, do we really want that numbers with decimal part are printed with `e` in `R` rather than with the decimal part?

proux01 replied:
Damn it, it was introduced in 8.10 and I'm the one to blame.

Printing in `R` could indeed be nicer. I'm adding that on top of my TODO list.

proux01 commented on interp/numTok.mli:
```diff
@@ -67,22 +85,38 @@ sig
+    (** [sprint] and [print] returns the numeral as it was parsed, for printing *)
+
+  val parse_string : string -> t option
+    (** Parse the string as a positive Coq numeral, if possible *)
```
s/positive //

herbelin replied:
Indeed.

proux01 commented on parsing/g_prim.mlg:
```diff
@@ -27,6 +27,13 @@ let my_int_of_string ?loc s =
+  | Some n -> n
+  | None ->
+    let pos = if ispos then "natural" else "integer" in
+    CErrors.user_err ?loc Pp.(str "This number is not a " ++ str pos ++ str " number.")
```
I would rather write "an integer", but my english might be wrong.

herbelin replied:
My English is also poor, but I agree with you.

proux01 commented:
> I'm unsure that having `_` is actually very intuitive and I'm wondering whether we should not be more restrictive. I don't know the general feeling.

I don't have any strong opinion. The feature seem less pervasive than I initially though. For instance, OCaml and Java both allow `_` everywhere, whereas C or Python 2 don't allow it at all. Python 3 doesn't allows successive `_` but single underscores are still accepted everywhere.

> Should we parse +42, or +3.5,

No opinion, but if it does not harm, certainly nice to have.

> Should I complete the API for consistency?

It's an internal API, probably no need to introduce too much dead code.

> Should I rename Signed/Unsigned into SignedDecimal/UnsignedDecimal

I personally wouldn't. We might for instance want to extend them to hexadecimal at some point in the future.

herbelin commented:
> No opinion, but if it does not harm, certainly nice to have.

I don't think it would be more difficult than parsing `-` (which is not easy)! It is more about whether this is intuitively expected.

> It's an internal API, probably no need to introduce too much dead code.

OK to me.

> >  Should I rename Signed/Unsigned into SignedDecimal/UnsignedDecimal
> 
> I personally wouldn't. 

OK to me.

> We might for instance want to extend them to hexadecimal at some point in the future.

Indeed a nice perspective.

SkySkimmer commented:
>OK for test suite

(the test suite is broken but not the part governed by the test suite codeowner team)

herbelin commented:
@coq/ci-maintainers: I've seen several time the GitLab pipeline failing without building nothing. Do you know the reason? What should I do?

ejgallego commented:
> What should I do?

See https://github.com/coq/coq/issues/11320 , you need to cancel a job and then restart it manually :/

herbelin commented:
If no more comments, I'm going to smash the commits.

proux01 commented:
Ok for me. You may want to add the above suggested test : https://github.com/coq/coq/pull/11703#discussion_r388760268

herbelin commented:
> You may want to add the above suggested test

Indeed. Sorry about missing them. Done, and squashed.

Note: I reverted the commit `Notation of the form "-x" should be written "- x"` as this does not fix all (factorization) issues with negative numbers; I will address these issues in another PR.

By the way, if ever you are interested to be maintainers for such kind of files, my understanding is that you can get in touch with @zimmi48.

proux01 commented on interp/numTok.mli:
```diff
@@ -1,18 +1,132 @@
+
+  val to_bigint_fraction : t -> Bigint.bigint * Bigint.bigint
+  val of_bigint_fraction : Bigint.bigint -> Bigint.bigint -> t
+    (** n and p such that the number is n/10^p *)
```
BTW, I wonder whether this isn't actually `n*10^p`, in which case the name might be a bit strange. But that's just a detail.

herbelin replied:
Yes, and I don't know why I started to be convinced it was a `/`. Thanks, it is not a detail after all. What about calling them `to_bigint_and_exponent` and `of_bigint_and_exponent` then?

proux01 replied:
sounds good

ppedrot commented:
@herbelin this has been sitting for a long time, I'll merge it when you rebase and after the CI.

herbelin commented:
I rebased. The VST failure seems specific to VST. The failures in Equations/Elpi/rewriter are due to the overlays of #11731 not being merged yet. Note the existence of an overlay for QuickChick.
